File: javaSorts/Heapsort.java
package javaSorts;

public class Heapsort {

    /**
     * Returns index multiplied by 2 and then incremented by 1
     * <p>
     * Running Time: O(1) = <b>O(1)</b>
     */
    private static int left(int index) {
        return (2 * index) + 1;
    }

    /**
     * Returns index multiplied by 2 and then incremented by 2
     * <p>
     * Running Time: O(1) = <b>O(1)</b>
     */
    private static int right(int index) {
        return (2 * index) + 2;
    }

    /**
     * Max heapify is used to maintain max-heap property.
     * <p>
     * Running Time: O(1) + O(1) + O(1) + O(1) + O(1) + O(1) + O(1) + O(1) + O(1) + O(1) + O(log n) = O(log n + 10) = <b>O(log n)</b>
     * <p>
     * log being base 2: log_{2}
     */
    private static void maxHeapify(int[] array, int index, int size) {
        int leftIndex = left(index);
        int rightIndex = right(index);
        int largest = index;

        if(leftIndex < size && array[leftIndex] > array[largest]) {
            largest = leftIndex;
        }
        
        if(rightIndex < size && array[rightIndex] > array[largest]) {
            largest = rightIndex;
        }

        if(largest != index) {
            HelperFunctions.swap(array, index, largest);
            maxHeapify(array, largest, size);
        }
    }

    /**
     * Build max heap builds max-heap from unordered array.
     * <p>
     * Running Time: O(1) + O(n / 2 - 1) + O(n log n / 2 - log n) = <b>O(n log n)</b> <p>
     * However operations are distributed in efficient way over diffrent tree depths leading to overall time complexity of: <p>
     * Running Time: <b>O(n)</b>
     */
    private static void buildMaxHeap(int[] array, int size) {
        int startingIndex = (size / 2) - 1;

        for (int index = startingIndex; index > -1; index--) {
            maxHeapify(array, index, size);
        }
    }

    private static void sort(int[] array, int size) {
        buildMaxHeap(array, size);
        int heapSize = size;

        for(int index = size - 1; index > 0; index--) {
            HelperFunctions.swap(array, 0, index);
            heapSize -= 1;
            maxHeapify(array, 0, heapSize);
        }
    }

    public static void sort(int[] array) {
        sort(array, array.length);
    }
}

File: python/heapsort.py
from helper_functions import swap


def __left(index:int):
    """
    Returns index multiplied by 2 and then incremented by 1

    Running Time: O(1) = <b>O(1)</b>
    """
    return (2 * index) + 1

def __right(index:int):
    """
    Returns index multiplied by 2 and then incremented by 2

    Running Time: O(1) = <b>O(1)</b>
    """
    return (2 * index) + 2

def __max_heapify(array:list, index:int, size:int):
    """
    Max heapify is used to maintain max-heap property.

    Running Time: O(1) + O(1) + O(1) + O(1) + O(1) + O(1) + O(1) + O(1) + O(1) + O(1) + O(log n) = O(log n + 10) = <b>O(log n)</b>

    log being base 2: log<sub>2</sub>
    """
    left_index = __left(index)
    right_index = __right(index)

    if left_index < size and array[left_index] > array[index]:
        largest = left_index
    else:
        largest = index
    
    if right_index < size and array[right_index] > array[largest]:
        largest = right_index
    
    if largest != index:
        swap(array, index, largest)
        __max_heapify(array, largest, size)

def __build_max_heap(array:list, size:int):
    """
    Build max heap builds max-heap from unordered array.

    Running Time: O(1) + O(n / 2 - 1) + O(n log n / 2 - log n) = <b>O(n log n)</b> \n
    However operations are distributed in efficient way over diffrent tree depths leading to overall time complexity of: \n
    Running Time: <b>O(n)</b>
    """
    starting_index = (size // 2) - 1

    for index in range(starting_index, -1, -1):
        __max_heapify(array, index, size)

def __heapsort(array:list, size:int):
    """
    Heapsort starts with building max heap then makes variable heap_size to be size of the heap. Then loop starts from size 
    decremented by 1 down to 2nd index, each time loop runs; 1st index is being swapped with loop index and heap_size is 
    decremented by 1, and max heapify is being called. After loop finishes array is sorted.

    Running Time: O(n) + O(1) + O(n - 2) + O(n - 2) + O(n - 2) + O(n log n - 2 log n) = O(n log n + 4n - 2 log n - 5) = <b>O(n log n)</b>
    """
    __build_max_heap(array, size)
    heap_size = size

    for index in range(size - 1, 0, -1):
        swap(array, 0, index)
        heap_size -= 1
        __max_heapify(array, 0, heap_size)

def heapsort(array:list):
    __heapsort(array, len(array))